### PR TITLE
Add snake to camel case fixer.

### PR DESCRIPTION
**Pull request by vmelnik-ukraine**
To avoid/fix snake case variables in the code base.

### CODE REVIEW

File: tests/Fixer/Naming/SnakeCaseToCamelCaseFixerTest.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+     */
+    public function provideFixCases()
+    {
+        return array(
+            // should not fix
+            array('<?php $variable = 1;'),
+            // should fix
+            array(
+                '<?php $camelCase = 1;',
+                '<?php $camel_case = 1;',
```
Comment (keradus) [inline tests/Fixer/Naming/SnakeCaseToCamelCaseFixerTest.php]:
```diff
 <?php
-$camel_case = 1;
-$camelCase = 1;
```

this changes code behaviour.

------

this does not
```diff
 <?php
 function foo ()
 {
-    $camel_case = 1;
+    $camelCase = 1;
 }
```

**Comment by keradus** [comment]
should be configurable:
- what: vars, properties, functions, methods, classes
- how: safe, all

like it's safe to change private methods/properties/non-global vars, but it's risky to change all, like global vars (test cases you have provided), protected methods, classes, ...

**Comment by SpacePossum** [comment]
>like it's safe to change private methods/properties/non-global vars

sadly its not, you can still use the name in a dynamic context;

```php
public function callPrivateMethod($snakeNameCase = 'a_b_c') {
    $this->$snakeNameCase;
}

private function a_b_c()
{
}

```

**Comment by julienfalque** [comment]
Changing the name of a private property/method can even break code that uses reflection.

File: src/Fixer/Naming/SnakeCaseToCamelCaseFixer.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+     * {@inheritdoc}
+     */
+    public function fix(\SplFileInfo $file, Tokens $tokens)
+    {
+        foreach ($tokens as $index => $token) {
+            if (!$token->isGivenKind(T_VARIABLE)) {
+                continue;
+            }
+
+            if (preg_match('/_/', $token->getContent())) {
```
Comment (julienfalque) [inline src/Fixer/Naming/SnakeCaseToCamelCaseFixer.php]: Not sure but maybe `false !== strpos($token->getContent(), '_')` would be more efficient.

File: tests/Fixer/Naming/SnakeCaseToCamelCaseFixerTest.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * @internal
+ */
+final class SnakeCaseToCamelCaseFixerTest extends AbstractFixerTestCase
+{
+    /**
+     * @dataProvider provideFixCases
+     *
+     * @param string $expected
+     * @param null   $input
```
Comment (julienfalque) [inline tests/Fixer/Naming/SnakeCaseToCamelCaseFixerTest.php]: `null|string`

**Comment by julienfalque** [comment]
We need more test cases: like `"$variable inside string"` for example.

**Comment by keradus** [comment]
> Also, some people prefer to avoid consecutive uppercase letters, e.g.: `getOrmInstance()` instead of `getORMInstance()`, should the fixer support this case?

Proper `camelCase` transformation for `to_JSON` is `toJson`, as `toJSON` would mean `to_j_s_o_n`. Discussed dozen of times on devs communities

**Comment by vmelnik-ukraine** [comment]
> Changing the name of a private property/method can also break code that uses reflection.

@julienfalque, seems like we'll have to mark the whole fixer as risky then.

> Like @keradus said, this fixer should also cover properties, methods, ... and be configurable.

@julienfalque, yes, I agree.

> What about supporting other casings like with constants, e.g. SOME_CONSTANT_VALUE instead of SomeConstantValue?

Shouldn't we have a dedicated fixer for that? I can't remember any other cases where I'd use this type of naming except with constants, so seems it's not directly related to snake vs. camel problem. Wdyt?

> Also, some people prefer to avoid consecutive uppercase letters, e.g.: getOrmInstance() instead of getORMInstance(), should the fixer support this case?

> Proper camelCase transformation for to_JSON is toJson, as toJSON would mean to_j_s_o_n. 

@keradus, @julienfalque if it's a common approach, then let's add a test for that

**Comment by keradus** [comment]
as a POC, when you will be happy about proposed solution include it in `.php_cs.dist` file as well, then we will see how it works with quite a piece of code.

**Comment by GrahamCampbell** [comment]
I'm 👎 on the idea. There are so many ways this can break code.

**Comment by keradus** [comment]
That's why I asked to add it to local config file. If author will manage to propose a fixer that would not broke PHP CS Fixer codebase, it may be good enought

**Comment by julienfalque** [comment]
> Shouldn't we have a dedicated fixer for that? I can't remember any other cases where I'd use this type of naming except with constants, so seems it's not directly related to snake vs. camel problem. Wdyt?

It's not a snake vs. camel case but it's still a naming style case that could be fixed. I won't be against extending the scope of the fixer here, maybe with configuration like:

```php
[
    'naming_styles' => [
        'constants' => 'uppercase_underscore',
        'variables' => 'camel_case',
        'functions' => 'snake_case',
        'methods' => 'camel_case',
        // ...
    ],
]
```

**Comment by keradus** [comment]
`'classes' => 'PascalCase'`

**Comment by GrahamCampbell** [comment]
Be careful with that. That may conflict with PSR0/4 fixers.

**Comment by GrahamCampbell** [comment]
Infact, it should only be allowed to be run if those fixers are off.

**Comment by keradus** [comment]
> That's why I asked to add it to local config file. If author will manage to propose a fixer that would not broke PHP CS Fixer codebase, it may be good enough

@vmelnik-ukraine, you have made nice POC for starting the work. We are aware that this fixer is not obvious and extra tricky. Are you taking the challenge to finish it making the state of fixer good enough to be run over codebase of PHP CS Fixer itself?

**Comment by vmelnik-ukraine** [comment]
@keradus yes, I'd like to give it a try.

**Comment by keradus** [comment]
Sadly, no changes here, thus I'm closing this PR.
Feel free to open it once more after you will try to apply the changes

**Pull request closed without merging**